lib/swaps/SwapClientManager.ts

```diff
@@ -0,0 +1,224 @@
+          lndConfig,
+          currency,
+          (SwapClientManager.wrapLndLogger(this.loggers.lnd, currency) as Logger),
+        );
+        this.swapClients.set(currency, lndClient);
+        initPromises.push(lndClient.init());
+      }
+    }
+    // setup raiden client and connect if configured
+    const raidenEnabled = !this.config.raiden.disable;
+    if (raidenEnabled) {
+      this.raidenClient = new RaidenClient(this.config.raiden, this.loggers.raiden);
+      initPromises.push(this.raidenClient.init());
+    }
+    // bind event listeners
+    this.bind();
+    await Promise.all(initPromises);
+    // associate swap clients with currencies managed by raiden client
+    if (raidenEnabled) {
+      for (const currency of this.raidenClient.tokenAddresses.keys()) {
+        this.swapClients.set(currency, this.raidenClient);
+      }
+    }
+  }
+
+  /**
+   * Gets a swap client instance.
+   * @param currency a currency that the swap client is linked to.
+   * @returns swap client instance upon success, throws otherwise.
+   */
+  public get(currency: string): BaseClient {
+    const swapClient = this.swapClients.get(currency);
+    if (!swapClient) {
+      throw errors.SWAP_CLIENT_NOT_FOUND(currency);
+    }
+    return swapClient;
+  }
+
+  /**
+   * Adds a new swap client and currency association.
+   * @param currency a currency that should be linked with a swap client.
+   * @returns Nothing upon success, throws otherwise.
+   */
+  public add(currency: Currency): void {
+    if (currency.swapClient === SwapClient.Raiden && currency.tokenAddress) {
+      this.swapClients.set(currency.id, this.raidenClient);
+      this.raidenClient.tokenAddresses.set(currency.id, currency.tokenAddress);
+    } else if (currency.swapClient === SwapClient.Lnd) {
+      // in case of lnd we check if the configuration includes swap client
+      // for the specified currency
+      let hasCurrency = false;
+      for (const lndCurrency in this.config.lnd) {
+        if (lndCurrency === currency.id) {
+          hasCurrency = true;
+        }
+      }
+      // adding a new lnd client at runtime is currently not supported
+      if (!hasCurrency) {
+        throw errors.SWAP_CLIENT_NOT_CONFIGURED(currency.id);
+      }
+    }
+  }
+
+  /**
+   * Removes a new swap client and currency association.
+   * @param currency a currency that should be unlinked from a swap client.
+   * @returns Nothing upon success, throws otherwise.
+   */
+  public remove(currencyId: string): void {
+    const swapClient = this.get(currencyId);
+    this.swapClients.delete(currencyId);
+    if (swapClient.type === SwapClient.Raiden) {
+      this.raidenClient.tokenAddresses.delete(currencyId);
+    }
+  }
+
+  /**
+   * Gets all lnd clients' pubKeys.
+   * @returns An object containing lnd public keys.
+   */
+  public getLndPubKeys(): LndPubKeys {
+    const lndPubKeys: LndPubKeys = {};
+    for (const [currency, swapClient] of this.swapClients.entries()) {
```

Comment:
What do you think about maintaining this list of pubkeys (similar to how we're maintaining the list of raiden currencies) rather than determining it dynamically when needed? I don't think it will be changing very often, so it might be ever so slightly clearer/more performant.

Reply:
I'm not completely sold on the idea because of the added overhead of making sure that list stays up to date with swap clients' values. I added a TODO in the code for further consideration.